File: rails-server-manager/app/models/user_group.rb
class UserGroup < ApplicationRecord
  # Associations
  has_many :user_group_memberships, dependent: :destroy
  has_many :users, through: :user_group_memberships
  has_many :projects, dependent: :nullify

  # Validations
  validates :name, presence: true, uniqueness: true

  # Methods
  def add_user(user, role: 'member')
    user_group_memberships.create(user: user, role: role)
  end

  def remove_user(user)
    user_group_memberships.find_by(user: user)&.destroy
  end

  def has_user?(user)
    users.include?(user)
  end

  def admin_users
    user_group_memberships.where(role: 'admin').map(&:user)
  end
end

File: laravel-server-manager/app/Models/UserGroup.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class UserGroup extends Model
{
    use HasFactory;

    protected $fillable = ['name', 'description'];

    // Relationships
    public function userGroupMemberships()
    {
        return $this->hasMany(UserGroupMembership::class);
    }

    public function users()
    {
        return $this->belongsToMany(User::class, 'user_group_memberships')
                    ->withPivot('role')
                    ->withTimestamps();
    }

    public function projects()
    {
        return $this->hasMany(Project::class);
    }

    // Helper methods
    public function addUser(User $user, string $role = 'member')
    {
        return $this->userGroupMemberships()->create([
            'user_id' => $user->id,
            'role' => $role,
        ]);
    }

    public function removeUser(User $user)
    {
        return $this->userGroupMemberships()->where('user_id', $user->id)->delete();
    }
}
